### PR TITLE
[FLINK-22276][runtime] Fixes concurrency bug in exception history

### PR DESCRIPTION
**Pull request by XComp**
## What is the purpose of the change

This commit fixes a case where multiple failures can occur close to each other. In that case, the `DefaultScheduler`'s [restart logic](https://github.com/apache/flink/blob/768c04a501602cf706cd56df1e258e5a292c8da4/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java#L258) competes for each of these failures. If multiple failures refer to the same `Execution`, it might be that the restart due to one failure handling cleans up the failure already. This leads to an `IllegalArgumentException` when archiving the next failure refering to the same `Execution` as the root cause. The issue was that the code relied on `ExecutionVertexID`s provided by the `FailureHandlingResult` instead of the actual `Executions`. 

## Brief change log

The new implementation relies on the `Executions` that were present when the failure was handled. Therefore, `FailureHandlingResultSnapshot` is introduced. It extracts the `Execution` information from the `ExecutionGraph` while the failure is processed.

Additionally, instead of accessing on `ExecutionVertex.getTaskNameWithSubtaskIndex()` to collect the task name, the new implementation relies on `Execution.getVertexWithAttempt()`. This enables us to solely rely on the `Execution` without this extra dependency on the `ExecutionVertex`.

The new implementation also removes the `add` method from `RootExceptionHistoryEntry`. This makes the instantiation cleaner. `ExceptionHistoryEntryExtractor` was replaced by the factory methods `RootExceptionHistoryEntry.fromFailureHandlingResultSnapshot` and `RootExceptionHistoryEntry.fromFailureHandlingResultSnapshot` as part of this effort.

## Verifying this change

* `DefaultSchedulerTest.testExceptionHistoryConcurrentRestart` is added to cover [FLINK-22276](https://issues.apache.org/jira/browse/FLINK-22276)
  * I needed to introduce a new `ScheduledExecutor` implementation for the test since I needed to change the order of the `ScheduledTasks` before actually triggering them (see `DefaultSchedulerTest.ReorganizableManuallyTriggeredScheduledExecutor`)
* `FailureHandlingResultSnapshotTest` was added to test the filtering of failed `Executions`
* `ExceptionHistoryEntryTest` and `RootExceptionHistoryEntryTest` were introduced to replace the test coverage which was provided by `ExceptionHistoryEntryExtractorTest` (except for the subtask ID tests that are not necessary anymore)
* `JobExceptionsHandlerTest` was only adapted to consider the refactorings as part of this PR

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by XComp** [comment]
I did a manual test checking the UI and the logs for the happy path.
Screenshot 2021-04-16 at 08 15 22

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:1185]: We don't have to change it right now but the test relies quite a bit on internal implementation details. To give you an example: If we decide to change how restarting tasks is internally handled, then we will break this test.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:636]: Here we are still relying on the `ExecutionVertices`. Shall we change it?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:42]: `@Nullable` is missing.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:86]: `getError` returns a field which is `@Nullable`. This violates the constructor.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:82]: I don't think that this is the most efficient way to look up the required information. If the EG is quite large and the failure region is small, then it will be quite costly. Maybe instead of passing in an `Iterable<ExecutionVertex>` we could pass in a lookup function `Function<ExecutionVertexID, ExecutionVertex>` which uses the `JobVertexID` and the `subtaskIndex` to look up the `ExecutionVertex`. This is not super important but if this is easy to change, it might save us some CPU cycles.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:144]: nit: Not the most efficient way to do it. Alternatively one could store `rootCauseExecution` and other failed `Executions` in separate fields.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:42]: Why do we store the `ExecutionAttemptID` instead of the `Execution` here?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:45]: Same here, why do we need to index the set of failed executions? Wouldn't `Set<Execution>` be good enough?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java:39]: Shall we make this a `Set`? This is not a strong argument but `Iterable` feels as if this class does not really own the `ExceptionHistoryEntries`, whereas if it is a `Set` or any other `Collection`, then it feels a bit different.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java:71]: Instead of doing this, could we call `fromGlobalFailure` or `internalCreationMethodSomething` because this part looks quite similar to what the `fromGlobalFailure` does. That way we avoid a bit of code duplication.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @XComp. The change looks good. I had a couple of comments which we should resolve before merging it. I like the thoroughness of your tests :-)

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java]: I added a workaround setting the exception explicitly to have a working version. The issue in general is again what's already covered by [FLINK-22060](https://issues.apache.org/jira/browse/FLINK-22060). Without the workaround, `JobMasterTest` is failing. AFAIR, it is a known issue that came up while doing the research on `FLINK-22060`. The `JobMasterTest` does not set the failure cause (because it didn't need to before).

The workaround fixes it for now and we should address the issue in a follow up ticket before releasing.

Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java]: > I added a workaround setting the exception explicitly to have a working version. The issue in general is again what's already covered by FLINK-22060. Without the workaround, JobMasterTest is failing. AFAIR, it is a known issue that came up while doing the research on FLINK-22060. The JobMasterTest does not set the failure cause (because it didn't need to before).
> The workaround fixes it for now and we should address the issue in a follow up ticket before releasing.

I re-iterated over the `@Nullable` constraints and updated the code accordingly. We're relying on a utility method `ErrorInfo.handleMissingThrowable` that addresses the issue of [FLINK-21376](https://issues.apache.org/jira/browse/FLINK-21376)/[FLINK-22060](https://issues.apache.org/jira/browse/FLINK-22060) for now.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: Would you have suggested to pick the `ScheduledTask` right after it was issued through `initiateFailure` and set the order of the `ScheduledTasks` explicitly?!

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java:164]: The test also passes w/o this wrapping.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:95]: Maybe add `@Nullable` annotation to `getError()` method as a hotfix.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:153]: I would suggest to return `Collections.unmodifiableSet(concurrentlyFailedExecutions)`.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @XComp. The changes look good to me. I had a last set of minor comments. Please take a look and resolve them. If you want, then you can also squash some of the commits.

Merging this PR once the last comments have been resolved and AZP gives green light.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java:164]: Good catch. I did some code changes afterwards that made the wrapping obsolete. 👍

**Comment by XComp** [comment]
Thanks for your review, @tillrohrmann . I addressed all your comments and squashed the commits.

**Comment by tillrohrmann** [comment]
Thanks @XComp. The test failure seems to be unrelated. Merging this PR now.

**Comment by XComp** [comment]
FYI: The failure might be related to [FLINK-22085](https://issues.apache.org/jira/browse/FLINK-22085). I added a comment to the ticket and re-opened it as it appears that we included the fix of `FLINK-22085` already in that build.

**Pull request closed without merging**
